controllers/clustertopology_controller_compute.go

```diff
@@ -140,6 +141,149 @@ func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clust
+// in the api server an error is thrown. If the control plane requires infrastructure as configured in the clusterClass
+// an error will be thrown if the control plane is found but with no infrastructure.
+func (r *ClusterTopologyReconciler) getControlPlaneState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (controlPlaneTopologyState, error) {
+	err := utilconversion.ConvertReferenceAPIContract(ctx, r.Client, r.restConfig, cluster.Spec.ControlPlaneRef)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrap(err, "control plane")
+	}
+	cp, err := external.Get(ctx, r.Client, cluster.Spec.ControlPlaneRef, cluster.Namespace)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrap(err, "control plane")
+	}
+	// Some control plane providers may not require infrastructure to run. This check returns early if the field
+	// indicating infrastructure is required is not found in the clusterClass of the given cluster
+	if class.Spec.ControlPlane.MachineInfrastructure == nil {
+		return controlPlaneTopologyState{
+			cp,
+			nil,
+		}, nil
+	}
+	cpInfraRef, err := getNestedRef(cp, "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "control plane infra ref")
+	}
+	err = utilconversion.ConvertReferenceAPIContract(ctx, r.Client, r.restConfig, cpInfraRef)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "control plane infrastructure reference not found")
+	}
+	infra, err := external.Get(ctx, r.UnstructuredCachingClient, cpInfraRef, cpInfraRef.Namespace)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "control plane infrastructure reference not found")
+	}
+	return controlPlaneTopologyState{
+		cp,
+		infra,
+	}, nil
+}
+
+// getCurrentMachineDeploymentState queries the apiserver for all MachineDeployments linked to a given cluster and managed
+// through a ClusterClass using labels. If it finds MachineDeployments it inspects them for infrastructure and bootstrap
+// references. Where these are not found the function will throw an error.
+func (r *ClusterTopologyReconciler) getCurrentMachineDeploymentState(ctx context.Context, cluster *clusterv1.Cluster) ([]machineDeploymentTopologyState, error) {
+	mDeploymentState := make([]machineDeploymentTopologyState, 0)
+
+	ms := &clusterv1.MachineDeploymentList{}
+	err := r.Client.List(ctx, ms, client.MatchingLabels{
+		clusterv1.ClusterLabelName: cluster.Name,
+		//TODO:Replace this with correct constant when merged
+		"cluster.x-k8s.io/topology": "",
+	})
+	if err != nil {
+		return mDeploymentState, errors.Wrap(err, "machine deployment")
+	}
+	for _, m := range ms.Items {
+		machine := m
+		infraRef := &m.Spec.Template.Spec.InfrastructureRef
+		if infraRef == nil {
+			return mDeploymentState, fmt.Errorf("%v", "no bootstrap reference in machine deployment")
```

Comment:
```suggestion
			return mDeploymentState, errors.Errorf(err, "machine deployment %s does not have a reference to a infrastructure machine template", md.Name)
```
Similar for the error messages below, let's use errors and surface MachineDeployment name